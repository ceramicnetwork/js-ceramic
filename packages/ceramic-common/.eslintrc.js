module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  plugins: [
    '@typescript-eslint',
    'jest'
  ],
  rules: {
    "@typescript-eslint/no-explicit-any": "off"
  },
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
  ],
  env: {
    "jest/globals": true
  }
};
