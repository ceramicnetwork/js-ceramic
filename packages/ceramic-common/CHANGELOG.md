# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 1.0.1-alpha.0 (2020-07-09)

**Note:** Version bump only for package @ceramicnetwork/ceramic-common





## 1.0.1-alpha.0 (2020-07-09)

**Note:** Version bump only for package @ceramicnetwork/ceramic-common
