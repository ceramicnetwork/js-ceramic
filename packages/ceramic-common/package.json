{
  "name": "@ceramicnetwork/ceramic-common",
  "version": "0.6.0",
  "description": "Ceramic common types and utilities",
  "keywords": [
    "ceramic",
    "types",
    "utilities"
  ],
  "author": "Janko Simonovic <simonovic86@gmail.com>",
  "homepage": "",
  "license": "(Apache-2.0 OR MIT)",
  "directories": {
    "lib": "lib"
  },
  "main": "lib/index.js",
  "files": [
    "lib"
  ],
  "scripts": {
    "test": "./node_modules/.bin/jest --coverage",
    "build": "../../node_modules/.bin/tsc -p tsconfig.json",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean",
    "lint": "./node_modules/.bin/eslint ./src --ext .js,.jsx,.ts,.tsx",
    "clean": "rm -rf ./lib"
  },
  "dependencies": {
    "@types/json-schema": "^7.0.5",
    "ajv": "^6.12.3",
    "cids": "^1.0.0",
    "dag-jose": "^0.2.0",
    "did-resolver": "^2.1.1",
    "lodash.clonedeep": "^4.5.0"
  },
  "devDependencies": {
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-decorators": "^7.10.1",
    "@babel/preset-env": "^7.8.4",
    "@babel/preset-typescript": "^7.9.0",
    "@types/encoding-down": "^5.0.0",
    "@types/events": "^3.0.0",
    "@types/lodash.clonedeep": "^4.5.6",
    "@types/node": "^13.13.15",
    "@types/prettier": "^2.0.2",
    "@typescript-eslint/eslint-plugin": "^2.19.0",
    "@typescript-eslint/parser": "^2.19.0",
    "babel-jest": "^25.1.0",
    "dids": "^0.5.0",
    "eslint": "^6.8.0",
    "eslint-plugin-jest": "^23.8.2",
    "identity-wallet": "^2.0.0-alpha.5",
    "jest": "^25.1.0",
    "jest-mock-extended": "^1.0.9",
    "js-sha256": "^0.9.0",
    "json-schema-to-typescript": "^9.1.1",
    "tmp-promise": "^2.0.2",
    "ts-node": "^9.0.0",
    "tsc-watch": "^4.2.9",
    "typescript-json-schema": "^0.42.0"
  },
  "jest": {
    "transformIgnorePatterns": [
      "/node_modules(?!/did-jwt)/"
    ]
  },
  "gitHead": "c58eb9659f5fdd976aca2db6465619c20ba30ff9"
}
