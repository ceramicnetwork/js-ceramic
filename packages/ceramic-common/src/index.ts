export * from './anchor-service'
export * from './ceramic-api'
export * from './context'
export * from './doctype'
export * from './utils/doctype-utils'
