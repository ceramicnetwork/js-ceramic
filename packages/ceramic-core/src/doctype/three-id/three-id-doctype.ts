import CID from 'cids'

import jsonpatch from 'fast-json-patch'

import {
    DocState, Doctype, InitOpts,
} from "../../doctype"
import { Context } from "../../context"
import User from "../../user"

const DOCTYPE = '3id'

/**
 * ThreeId doctype parameters
 */
export class ThreeIdParams {
    content: object;
    owners?: Array<string>;
}

/**
 * ThreeId doctype implementation
 */
export class ThreeIdDoctype implements Doctype {
    id: string;
    content: object
    doctype: string
    head: CID
    owners: string[]
    state: DocState

    /**
     * Change existing ThreeId doctype
     * @param params - Change parameters
     * @param context - Ceramic context
     */
    async change(params: ThreeIdParams, context: Context): Promise<ThreeIdDoctype> {
        return ThreeIdDoctype.change(this, params, context)
    }

    /**
     * Creates ThreeId doctype instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async create(params: ThreeIdParams, context: Context, opts?: InitOpts): Promise<ThreeIdDoctype> {
        const { content, owners } = params
        if (!owners) {
            throw new Error('The owner of the 3ID needs to be specified')
        }

        const record = await ThreeIdDoctype._makeGenesis(content, owners)
        const threeIdDocType = await context.api.createFromGenesis<ThreeIdDoctype>(record, opts)
        return Promise.resolve(threeIdDocType)
    }

    /**
     * Makes genesis record
     * @param content - Content
     * @param owners - Owners
     */
    static async _makeGenesis(content: any, owners?: Array<string>): Promise<any> {
        if (!owners) {
            throw new Error('The owner of the 3ID needs to be specified')
        }
        return {
            doctype: DOCTYPE, owners, content
        }
    }

    /**
     * Change existing ThreeId doctype
     * @param doctype - Doctype instance
     * @param params - Change parameters
     * @param context - Ceramic context
     */
    static async change(doctype: Doctype, params: ThreeIdParams, context: Context): Promise<ThreeIdDoctype> {
        const { content, owners } = params

        const updateRecord = ThreeIdDoctype._makeRecord(doctype, context.user, content, owners)
        return await context.api.applyRecord(doctype.id, updateRecord)
    }

    /**
     * Creates change record
     * @param doctype - Doctype instance
     * @param user - User instance
     * @param newContent - New context
     * @param newOwners - New owners
     * @private
     */
    static async _makeRecord(doctype: Doctype, user: User, newContent: any, newOwners?: string[]): Promise<any> {
        if (user == null) {
            throw new Error('No user authenticated')
        }

        if (typeof newContent == null) {
            newContent = doctype.state.content
        }

        const patch = jsonpatch.compare(doctype.state.content, newContent)
        const record: any = { owners: newOwners, content: patch, prev: doctype.head, id: doctype.state.log[0] }
        // TODO - use the dag-jose library for properly encoded signed records
        record.iss = user.DID
        // convert CID to string for signing
        const tmpPrev = record.prev
        const tmpId = record.id
        record.prev = { '/': tmpPrev.toString() }
        record.id = { '/': tmpId.toString() }
        const jwt = await user.sign(record, { useMgmt: true})
        const [header, payload, signature] = jwt.split('.') // eslint-disable-line @typescript-eslint/no-unused-vars
        record.prev = tmpPrev
        record.id = tmpId

        return { ...record, header, signature }
    }

}
