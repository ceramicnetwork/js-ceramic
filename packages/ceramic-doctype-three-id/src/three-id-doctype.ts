import jsonpatch from 'fast-json-patch'

import {
    Doctype,
    DoctypeConstructor,
    DoctypeStatic,
    DocOpts,
    DocParams
} from "@ceramicnetwork/ceramic-common"

import { DID } from 'dids'
import { Context } from "@ceramicnetwork/ceramic-common"

const DOCTYPE = '3id'

/**
 * ThreeId doctype parameters
 */
export interface ThreeIdParams extends DocParams {
    content?: object;
}

/**
 * ThreeId doctype implementation
 */
@DoctypeStatic<DoctypeConstructor<ThreeIdDoctype>>()
export class ThreeIdDoctype extends Doctype {
    /**
     * Change existing ThreeId doctype
     * @param params - Change parameters
     */
    async change(params: ThreeIdParams): Promise<void> {
        const { content, metadata } = params

        const updateRecord = await ThreeIdDoctype._makeRecord(this, this.context.user, content, metadata?.owners, metadata?.schema)
        const updated = await this.context.api.applyRecord(this.id, updateRecord)
        this.state = updated.state
    }

    /**
     * Creates ThreeId doctype instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async create(params: ThreeIdParams, context: Context, opts?: DocOpts): Promise<ThreeIdDoctype> {
        const { content, metadata } = params
        if (!metadata?.owners) {
            throw new Error('The owner of the 3ID needs to be specified')
        }

        const record = await ThreeIdDoctype.makeGenesis({ content, metadata })
        return context.api.createDocumentFromGenesis<ThreeIdDoctype>(record, opts)
    }

    /**
     * Creates genesis record
     * @param params - Create parameters
     */
    static async makeGenesis(params: Record<string, any>): Promise<Record<string, any>> {
        const { content, metadata } = params

        if (!metadata) {
            throw new Error('Metadata needs to be specified')
        }

        if (!metadata.owners) {
            throw new Error('The owner of the 3ID needs to be specified')
        }
        return {
            doctype: DOCTYPE, header: metadata, data: content
        }
    }

    /**
     * Creates change record
     * @param doctype - Doctype instance
     * @param user - User instance
     * @param newContent - New context
     * @param newOwners - New owners
     * @param newSchema - New schema
     * @private
     */
    static async _makeRecord(doctype: Doctype, user: DID, newContent: any, newOwners: string[] = null, newSchema: string = null): Promise<any> {
        if (user == null || !user.authenticated) {
            throw new Error('No user authenticated')
        }

        if (typeof newContent == null) {
            newContent = doctype.state.content
        }

        const patch = jsonpatch.compare(doctype.state.content, newContent)
        const header = doctype.metadata
        if (newOwners) {
            header.owners = newOwners
        }
        if (newSchema) {
            header.schema = newSchema
        }
        const record: any = { header, data: patch, prev: doctype.head, id: doctype.state.log[0] }
        // TODO - use the dag-jose library for properly encoded signed records
        // convert CID to string for signing
        const tmpPrev = record.prev
        const tmpId = record.id
        record.prev = { '/': tmpPrev.toString() }
        record.id = { '/': tmpId.toString() }

        const jws = await user.createJWS(record)
        const [signedHeader, payload, signature] = jws.split('.') // eslint-disable-line
        // @typescript-eslint/no-unused-vars
        record.prev = tmpPrev
        record.id = tmpId

        return { ...record, signedHeader, signature }
    }

}
