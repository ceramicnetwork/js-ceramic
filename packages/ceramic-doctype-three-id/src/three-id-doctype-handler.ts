import CID from 'cids'
import base64url from 'base64url'
import cloneDeep from 'lodash.clonedeep'
import stringify from 'fast-json-stable-stringify'

import jsonpatch from 'fast-json-patch'
import * as didJwt from 'did-jwt'
import { ThreeIdDoctype, ThreeIdParams } from "./three-id-doctype"
import {
    AnchorProof, AnchorRecord, AnchorStatus, DocState, DoctypeConstructor, DoctypeHandler, DocOpts, SignatureStatus
} from "@ceramicnetwork/ceramic-common"
import { Context } from "@ceramicnetwork/ceramic-common"

const DOCTYPE = '3id'

/**
 * ThreeId doctype handler implementation
 */
export class ThreeIdDoctypeHandler implements DoctypeHandler<ThreeIdDoctype> {
    /**
     * Gets doctype name
     */
    get name(): string {
        return DOCTYPE
    }

    /**
     * Gets doctype class
     */
    get doctype(): DoctypeConstructor<ThreeIdDoctype> {
        return ThreeIdDoctype
    }

    /**
     * Creates ThreeId doctype instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    async create(params: ThreeIdParams, context: Context, opts?: DocOpts): Promise<ThreeIdDoctype> {
        return ThreeIdDoctype.create(params, context, opts);
    }

    /**
     * Applies record (genesis|signed|anchor)
     * @param record - Record to be applied
     * @param cid - Record CID
     * @param context - Ceramic context
     * @param state - Document state
     */
    async applyRecord(record: any, cid: CID, context: Context, state?: DocState): Promise<DocState> {
        if (state == null) {
            // apply genesis
            return this._applyGenesis(record, cid)
        }

        if (record.proof) {
            const proofRecord = (await context.ipfs.dag.get(record.proof)).value;
            return this._applyAnchor(record, proofRecord, cid, state);
        }

        return this._applySigned(record, cid, state, context);
    }

    /**
     * Applies genesis record
     * @param record - Genesis record
     * @param cid - Genesis record CID
     * @private
     */
    async _applyGenesis(record: any, cid: CID): Promise<DocState> {
        if (record.doctype === DOCTYPE) {
            return {
                doctype: DOCTYPE,
                metadata: record.header,
                content: record.data,
                next: {
                    content: null,
                },
                signature: SignatureStatus.GENESIS,
                anchorStatus: AnchorStatus.NOT_REQUESTED,
                log: [cid]
            }
        } else if (record['@context'] === "https://w3id.org/did/v1") {
            const managementKey = record.publicKey.find((pk: { id: string }) => pk.id === 'did:3:GENESIS#managementKey').ethereumAddress
            const signingKey = record.publicKey.find((pk: { id: string }) => pk.id === 'did:3:GENESIS#signingKey').publicKeyHex
            const encryptionKey = record.publicKey.find((pk: { id: string }) => pk.id === 'did:3:GENESIS#encryptionKey').publicKeyBase64
            return {
                doctype: DOCTYPE,
                metadata: {
                    owners: [managementKey],
                },
                content: {
                    publicKeys: {
                        signing: signingKey, encryption: encryptionKey
                    }
                },
                next: {
                    content: null,
                },
                signature: SignatureStatus.GENESIS,
                anchorStatus: AnchorStatus.NOT_REQUESTED,
                log: [cid]
            }
        }
        // TODO - verify genesis record
    }

    /**
     * Applies signed record
     * @param record - Signed record
     * @param cid - Signed record CID
     * @param state - Document state
     * @param context - Ceramic context
     * @private
     */
    async _applySigned(record: any, cid: CID, state: DocState, context: Context): Promise<DocState> {
        if (!record.id.equals(state.log[0])) {
            throw new Error(`Invalid docId ${record.id}, expected ${state.log[0]}`)
        }
        const cloned = cloneDeep(record)

        const { signedHeader, signature } = cloned
        const payload = base64url.encode(stringify({
            doctype: cloned.doctype,
            data: cloned.data,
            header: cloned.header,
            unique: cloned.unique || undefined,
            prev: { '/': cloned.prev.toString() },
            id: { '/': cloned.id.toString() },
        }))

        const jws = [signedHeader, payload, signature].join('.')

        const decodedHeader = JSON.parse(base64url.decode(signedHeader))
        const { kid } = decodedHeader
        const { publicKey } = await context.resolver.resolve(kid)
        try {
            await this.verifyJWS(jws, publicKey)
        } catch (e) {
            throw new Error('Invalid signature for signed record. ' + e)
        }

        state.log.push(cid)
        return {
            ...state,
            signature: SignatureStatus.SIGNED,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            next: {
                owners: record.header.owners,
                content: jsonpatch.applyPatch(state.content, record.data).newDocument,
            },
        }
    }

    /**
     * Verifies JWS token
     * @param jwt - JWS token
     * @param pubkeys - public key(s)
     */
    async verifyJWS(jwt: string, pubkeys: any): Promise<void> {
        await didJwt.verifyJWS(jwt, pubkeys)
    }

    /**
     * Applies anchor record
     * @param record - Anchor record
     * @param proof - Anchor proof record
     * @param cid - Anchor record CID
     * @param state - Document state
     * @private
     */
    async _applyAnchor(record: AnchorRecord, proof: AnchorProof, cid: CID, state: DocState): Promise<DocState> {
        state.log.push(cid)
        let content = state.content
        if (state.next?.content) {
            content = state.next.content
            delete state.next.content
        }
        let owners = state.metadata?.owners
        if (state.next?.owners) {
            owners = state.next.owners
            delete state.next.owners
        }
        return {
            ...state,
            metadata: {
                owners
            },
            content,
            anchorStatus: AnchorStatus.ANCHORED,
            anchorProof: proof,
        }
    }

}
