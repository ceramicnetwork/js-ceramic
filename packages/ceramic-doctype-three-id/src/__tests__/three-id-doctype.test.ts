import CID from 'cids'

import { Resolver } from "did-resolver"
import ThreeIdResolver from '@ceramicnetwork/3id-did-resolver'

import { DID } from 'dids'
import { ThreeIdDoctypeHandler } from "../three-id-doctype-handler"
import { ThreeIdDoctype } from "../three-id-doctype"

import { Context } from "@ceramicnetwork/ceramic-common"
jest.mock('did-jwt', () => ({
  // TODO - We should test for when this function throws as well
  verifyJWS: (): void => { return }
}))

const cloneDeep = require('lodash.clonedeep') // eslint-disable-line @typescript-eslint/no-var-requires
const { sha256 } = require('js-sha256') // eslint-disable-line @typescript-eslint/no-var-requires
const hash = (data: string): CID => new CID(1, 'sha2-256', Buffer.from('1220' + sha256(data), 'hex'))

const FAKE_CID_1 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = new CID('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_3 = new CID('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_4 = new CID('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlsu')

const RECORDS = {
  genesis: { doctype: '3id', header: { owners: [ '0x123' ] }, data: { publicKeys: { test: '0xabc' } } },
  r1: {
    desiredContent: { publicKeys: { test: '0xabc' }, other: 'data' },
    record: { data: [ { op: 'add', path: '/other', value: 'data' } ], header: { owners: ['0x123'] }, id: FAKE_CID_1, prev: FAKE_CID_1, signedHeader: 'eyJraWQiOiJkaWQ6MzpiYWZ5YXNkZmFzZGY_dmVyc2lvbj0wI3NpZ25pbmciLCJhbGciOiJFUzI1NksifQ', signature: 'cccc' }
  },
  r2: { record: { proof: FAKE_CID_4 } },
  proof: {
    value: {
      blockNumber: 123456
    }
  }
}

describe('ThreeIdHandler', () => {
  let user: DID, threeIdDoctypeHandler: ThreeIdDoctypeHandler, context: Context

  beforeAll(() => {
    user = new DID()
    user.createJWS = jest.fn(async () => {
      // fake jws
      return 'eyJraWQiOiJkaWQ6MzpiYWZ5YXNkZmFzZGY_dmVyc2lvbj0wI3NpZ25pbmciLCJhbGciOiJFUzI1NksifQ.bbbb.cccc'
    })
    user._id = 'did:3:bafyuser'

    const recs: Record<string, any> = {}
    const ipfs = {
      dag: {
        put(rec: any, cid?: CID): any {
          if (cid) {
            recs[cid.toString()] = rec
            return cid
          }
          // stringify as a way of doing deep copy
          const clone = cloneDeep(rec)
          const c = hash(JSON.stringify(clone))
          recs[c.toString()] = clone
          return c
        },
        get(cid: any): any {
          return recs[cid.toString()]
        }
      }
    }

    const threeIdResolver = ThreeIdResolver.getResolver({
      loadDocument: (): any => {
        return Promise.resolve({
          content: {
            "publicKeys": {
              "signing": "zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
              "encryption": "z6LSfQabSbJzX8WAm1qdQcHCHTzVv8a2u6F7kmzdodfvUCo9"
            }
          }
        })
      }
    })

    context = {
      ipfs,
      resolver: new Resolver({
        ...threeIdResolver
      }),
      anchorService: null,
    }
  })

  beforeEach(() => {
    threeIdDoctypeHandler = new ThreeIdDoctypeHandler()
  })

  it('is constructed correctly', async () => {
    expect(threeIdDoctypeHandler.name).toEqual('3id')
  })

  it('makes genesis record correctly', async () => {
    await expect(ThreeIdDoctype.makeGenesis({ content: RECORDS.genesis.data })).rejects.toThrow(/Metadata needs/)
    const record = await ThreeIdDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: RECORDS.genesis.header })
    expect(record).toEqual(RECORDS.genesis)
  })

  it('applies genesis record correctly', async () => {
    await context.ipfs.dag.put(RECORDS.genesis, FAKE_CID_1)

    const state = await threeIdDoctypeHandler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    expect(state).toMatchSnapshot()
  })

  it('makes signed record correctly', async () => {
    await context.ipfs.dag.put(RECORDS.genesis, FAKE_CID_1)

    const state = await threeIdDoctypeHandler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    const doctype = new ThreeIdDoctype(state, context)

    await expect(ThreeIdDoctype._makeRecord(doctype, null, RECORDS.r1.desiredContent)).rejects.toThrow(/No user/)
    const record = await ThreeIdDoctype._makeRecord(doctype, user, RECORDS.r1.desiredContent)
    expect(record).toEqual(RECORDS.r1.record)
  })

  it('applies signed record correctly', async () => {
    await context.ipfs.dag.put(RECORDS.genesis, FAKE_CID_1)
    await context.ipfs.dag.put(RECORDS.r1.record, FAKE_CID_2)

    let state = await threeIdDoctypeHandler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    state = await threeIdDoctypeHandler.applyRecord(RECORDS.r1.record, FAKE_CID_2, context, state)
    expect(state).toMatchSnapshot()
  })

  it('applies anchor record correctly', async () => {
    await context.ipfs.dag.put(RECORDS.genesis, FAKE_CID_1)
    await context.ipfs.dag.put(RECORDS.r1.record, FAKE_CID_2)
    await context.ipfs.dag.put(RECORDS.r2.record, FAKE_CID_3)
    await context.ipfs.dag.put(RECORDS.proof, FAKE_CID_4)

    let state = await threeIdDoctypeHandler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    state = await threeIdDoctypeHandler.applyRecord(RECORDS.r1.record, FAKE_CID_2, context, state)
    state = await threeIdDoctypeHandler.applyRecord(RECORDS.r2.record, FAKE_CID_3, context, state)
    expect(state).toMatchSnapshot()
  })
})
