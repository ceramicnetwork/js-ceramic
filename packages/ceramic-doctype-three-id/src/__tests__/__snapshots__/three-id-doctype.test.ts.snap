// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeIdHandler applies anchor record correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
  },
  "anchorStatus": 3,
  "content": Object {
    "other": "data",
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "3id",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          119,
          169,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "metadata": Object {
    "owners": Array [
      "0x123",
    ],
  },
  "next": Object {},
  "signature": 2,
}
`;

exports[`ThreeIdHandler applies genesis record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "3id",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "metadata": Object {
    "owners": Array [
      "0x123",
    ],
  },
  "next": Object {
    "content": null,
  },
  "signature": 0,
}
`;

exports[`ThreeIdHandler applies signed record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "other": "data",
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "3id",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "metadata": Object {
    "owners": Array [
      "0x123",
    ],
  },
  "next": Object {
    "content": Object {
      "other": "data",
      "publicKeys": Object {
        "test": "0xabc",
      },
    },
    "owners": Array [
      "0x123",
    ],
  },
  "signature": 2,
}
`;
